### ISSUE TITLE
Agent should implement Circuit-Breaker Pattern

### ISSUE THREAD

**Issue by cagatatu** [elastic/apm-agent-dotnet#433]
**Is your feature request related to a problem? Please describe.**
The agent has a performance problem due to the HTTP queuing when the APM server or Elasticsearch is down. This message appears in logs one after another: {PayloadSenderV2} Failed sending event. queue is full

**Describe the solution you'd like**
I think it would be nice for the agent to implement Circuit-Breaker pattern described here: https://docs.microsoft.com/en-us/azure/architecture/patterns/circuit-breaker.

**Describe alternatives you've considered**
We had to manually disable the agent in our production environment.

**Additional context**
It is worth mentioning that there is no problem when the APM server and Elasticsearch are up and running. Also, when we had the problem, our application receives approximately 1000 requests per second.

**Comment by SergeyKleyman** [comment]
Hi @cagatatu. First of all thank you for trying out the agent and opening an issue to improve it.
Could you please clarify how "The agent has a performance problem" expresses itself? Does agent uses a lot of memory and/or CPU?

Regarding the solution you have proposed, using Circuit-Breaker Pattern - other Elastic APM agents (Java, Go, etc.) implement a back-off approach described at https://github.com/elastic/apm/blob/master/docs/agent-development.md#Transport-errors
and we have plans to implement it in Elastic APM .NET agent as well (https://github.com/elastic/apm-agent-dotnet/issues/141) - do you think it will address your use case?

**Comment by cagatatu** [comment]
@SergeyKleyman Thanks for the swift response.

Unfortunately, I cannot provide any more metrics for the problem. I can say it is not a CPU or RAM problem; we did not see any raise in them. I suspect the problem was related to IO threads, but I cannot be certain because we removed the APM agent from our application as soon as we can to mitigate the problem and cannot do further analysis.

I feel that the back-off approach will solve a lot of problems and it will exactly address our use case. That and the ability to activate/deactivate the agent completely with configuration (I've seen that you marked that one with 1.1 milestone on #122) will be a perfect couple for production use.

I guess this issue can be closed as it is so similar to #141 as you suggested.
